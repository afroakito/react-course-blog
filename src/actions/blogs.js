export const addBlog = (blog) => ({
    type: 'ADD_BLOG',
    blog,
});

export const editBlog = (id, updates) => ({
    type: 'EDIT_BLOG',
    id,
    updates
});