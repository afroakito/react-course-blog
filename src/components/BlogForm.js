import React from 'react';
import uuid from 'uuid';

export default class BlogForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            title: props.blog ? props.blog.title : '',
            body: props.blog ? props.blog.body : '',
            error: ''
        };
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.title || !this.state.body) {
            const error = 'Please provide title and body.';
            this.setState(() => ({ error }));
        } else {
            this.setState(() => ({ error: '' }));

            this.props.onSubmit({
                title: this.state.title,
                body: this.state.body
            });
        };
    };
    onTitleChange = (e) => {
        const title = e.target.value;

        this.setState(() => ({ title }));
    };
    onBodyChange = (e) => {
        const body = e.target.value;

        this.setState(() => ({ body }));
    };
    render () {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                <div>
                    {!!this.state.error && <p className="form_error">{this.state.error}</p>}
                </div>
                <input
                    type="text"
                    placeholder="Post Title"
                    autoFocus
                    className="text-input"
                    value={this.state.title}
                    onChange={this.onTitleChange}
                />
                <textarea
                    placeholder="Post Body"
                    className="textarea"
                    value={this.state.body}
                    onChange={this.onBodyChange}
                >
                </textarea>
                <div>
                    <button className="button">Save Post</button>
                </div>
            </form>
        )
    }
};