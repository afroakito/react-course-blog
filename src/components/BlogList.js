import React from 'react';
import { connect } from 'react-redux';
import BlogListItem from './BlogListItem';

export const BlogList = (props) => {
    console.log(props.blogs);
    return (
        <div className="content-container">
            <div className="list-header">
                <div>Blog Post</div>
            </div>
            <div className="list-body">
                {
                    props.blogs.map((blog) => {
                        return <BlogListItem key={blog.id} {...blog} />;
                    })
                }
            </div>
        </div>
    );
}

const mapStateToProps = (state) => {
    return {
        blogs: state.blogs
    };
};

export default connect(mapStateToProps)(BlogList);