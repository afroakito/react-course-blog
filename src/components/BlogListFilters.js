import React from 'react';
import { Link } from 'react-router-dom';

export const BlogListFilters = () => {
    return (
        <div className="content-container">
            <div className="header__filters">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type="text"
                            placeholder="Search blogs"
                            className="text-input"
                        />
                    </div>
                    <div className="input-group__item">
                        <select className="select">
                            <option value="byTitle">By title</option>
                            <option value="byDate">By date</option>
                        </select>
                    </div>
                </div>
                <Link className="button" to="/create">Add Blog</Link>
            </div>
        </div>
    );
};