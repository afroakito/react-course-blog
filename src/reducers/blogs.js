const blogsReducerDefaultState = [];

export default (state = blogsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_BLOG':
            return [
                ...state,
                action.blog,
            ];
        case 'EDIT_BLOG':
            return state.map((blog) => {
                if (blog.id === action.id) {
                    return {
                        ...blog,
                        ...action.updates
                    };
                } else {
                    return blog;
                }
            });
        case 'REMOVE_BLOG':
            return state.filter(({ id }) => {
                return id !== action.id
            });
        default:
            return state;
    }
};

// blog object image
// blog = {
//     title: '',
//     body: ''
// }