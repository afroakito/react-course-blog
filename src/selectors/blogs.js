// Get visible blogs

export default (blogs, { text }) => {
    return blogs.filter((blog) => {
        const textMatch = blog.title.toLowerCase().includes(text.toLowerCase());

        return textMatch;
    });
};