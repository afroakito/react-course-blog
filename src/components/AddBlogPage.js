import React from 'react';
import { connect } from 'react-redux';
import BlogForm from './BlogForm';
import { startAddBlog } from '../actions/blogs';

export class AddBlogPage extends React.Component {
    onSubmit = (blog) => {
        this.props.startAddBlog(blog);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <p>This is add blog page</p>
                <BlogForm onSubmit={this.onSubmit} />
            </div>
        );
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        startAddBlog: (blog) => dispatch(startAddBlog(blog))
    }
};

export default connect(undefined, mapDispatchToProps)(AddBlogPage);