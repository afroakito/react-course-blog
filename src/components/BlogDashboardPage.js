import React from 'react';
import BlogList from './BlogList';
import BlogListFilters from './BlogListFilters';

const BlogDashboardPage = () => (
  <div>
    <BlogListFilters />
    <BlogList />
  </div>
);

export default BlogDashboardPage;
