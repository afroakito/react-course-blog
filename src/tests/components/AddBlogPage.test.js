import React from 'react';
import { shallow } from 'enzyme';
import { AddBlogPage } from '../../components/AddBlogPage';
import blogs from '../fixtures/blogs';

test('should render AddBlogPage correctly', () => {
    const wrapper = shallow(<AddBlogPage />);
    expect(wrapper).toMatchSnapshot();
})

test('should handle onSubmit', () => {
    const startAddBlog = jest.fn();
    const history = { push: jest.fn() };

    const wrapper = shallow(<AddBlogPage startAddBlog={startAddBlog} history={history} />);
    wrapper.find('BlogForm').prop('onSubmit')(blogs[1]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startAddBlog).toHaveBeenLastCalledWith(blogs[1]);
});