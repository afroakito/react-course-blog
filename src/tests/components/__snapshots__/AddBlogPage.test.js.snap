// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddBlogPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Add Blog
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <BlogForm
      onSubmit={[Function]}
    />
  </div>
</div>
`;
